src/gpt.py

```diff
@@ -0,0 +1,312 @@
+    # Load additional context from history
+    word_triggers = set()
+    user_triggers = set()
+    for content, _ in history:
+        for word in word_trigger_pattern.findall(content):
+            word_triggers.add(word)
+        for user in user_trigger_pattern.findall(content):
+            user_triggers.add(user)
+    # Return context about users + word triggers from history
+    return [f"{user} info: {user_info_map[user]}" for user in user_triggers] + [
+        context_data["word_triggers"][word] for word in word_triggers
+    ]
+
+
+# Make an api query
+async def get_response_text(message: Message) -> Optional[str]:
+    context = await get_message_context(message)
+    if random.random() < 0.05 or disallowed_message(message):
+        # If message is disallowed (or the user is unlucky), pass a special instruction
+
+        # Get user info since we're not passing history which would trigger it
+        user = get_name(message.author)
+        context.append(f"{user}: {user_info_map[user.lower()]}")
+        # Pass special instruction
+        context.append(context_data["banned_phrase_instruction"])
+        # Make up empty history so she understands the format
+        history = [(f"{self_user.name}:", True), (f"{user}:", False)]
+    else:
+        # Load history with 512 token limit and 5 speaking turn limit
+        history = await fetch_history(512, 5, message)
+        # We couldn't fit even a single message in history
+        if not history:
+            message.reply("I'm not reading all that.")
+            return
+
+        context += get_history_context(history)
+
+    # Send data to API
+    data = []
+    data.append({"role": "system", "content": "\n".join(context)})
+
+    for msg, is_self in reversed(history):
+        role = "assistant" if is_self else "user"
+        data.append({"role": role, "content": msg})
+
+    response = await query_api(data)
+    if response:
+        # Remove "Busty: " prefix
+        prefix = f"{get_name(self_user)}: "
+        if response.startswith(prefix):
+            response = response[len(prefix) :]
+
+        return response
+    else:
+        await message.reply("busy rn")
+    return None
+
+
+async def respond(message: Message):
+    # React with a "stop" hand, we're responding to someone else
+    if gpt_lock.locked() or not context_data:
+        raised_hand_emoji = "\N{RAISED HAND}\U0001F3FF"
+        await message.add_reaction(raised_hand_emoji)
+        return
+    # Respond to message
+    async with gpt_lock:
+        async with message.channel.typing():
+            response = await get_response_text(message)
+        if response:
+            response_split = [
+                response[i : i + config.MESSAGE_LIMIT]
+                for i in range(0, len(response), config.MESSAGE_LIMIT)
+            ]
+            # If user's message is most recent message in channel, send
+            # If others have been sent in the meantime, reply
```

Comment:
Just to make this read a bit clearer (ChatGPT was not used to generate this diff):

```suggestion
            # Reply to the message if it's not the most recent on in the chat history
```